daemon/containerd/image_pull.go

```diff
@@ -124,7 +125,12 @@ func (i *ImageService) PullImage(ctx context.Context, ref reference.Named, platf
 	img, err := i.client.Pull(ctx, ref.String(), opts...)
 	if err != nil {
 		if errors.Is(err, docker.ErrInvalidAuthorization) {
-			return errdefs.NotFound(fmt.Errorf("pull access denied for %s, repository does not exist or may require 'docker login'", reference.FamiliarName(ref)))
+			outErr := fmt.Errorf("pull access denied for %s, repository does not exist or may require 'docker login'", reference.FamiliarName(ref))
+
+			if strings.Contains(err.Error(), "no basic auth credentials") {
```

Comment:
Oh; I guess I get it now; this is the part that's checked for in `retryOnError`, is that the issue? https://github.com/moby/moby/blob/f6576e9a741d8ee9152a224d2ef342decdcab93c/distribution/errors.go#L166-L171

Should we match that error instead of a string here? (`auth.ErrNoBasicAuthCredentials`). Also wondering if we could come with a different approach to prevent  retries (should we add more things to `retryOnError` ?

Reply:
Containerd doesn't return `ErrNoBasicAuthCredentials` but a `ErrInvalidAuthorization` in the case where no basic auth is present

Reply:
No, it has nothing to do with `auth.ErrNoBasicAuthCredentials`. It's an error returned from containerd: https://github.com/dmcgowan/containerd/blob/51c9ffe468f8752d851ab6efda61a2f55026edfc/remotes/docker/authorizer.go#L190

The retry code from distribution isn't used at all with the containerd pull.

Reply:
Ah, right, but isn't checking for the `ErrInvalidAuthorization` enough for this, or are there cases where `ErrInvalidAuthorization` is returned and we should _not_ decorate the error with `pull access denied for %s, repository does not exist or may require 'docker login'`?

If there is no _technical_ reason for the `no basic auth credentials`, to be required, and if it's only to make the tests happy, we should tread with care.

👉 Basically, I want to make sure we're not codifying this string to become an API contract if there is no need; of course changing strings _MAY_ still get some people complaining ("I was checking for that string!"), but we could still look at that if we get reports.
 

Looking at tests checking for this string, I found `TestLogoutWithExternalAuth` and/or `TestPushNoCredentialsNoRetry`; both of those have already been updated a few times (see https://github.com/moby/moby/commit/745892a7b271cf0f1770a8ec3698aaf61573e5f2, https://github.com/moby/moby/commit/8d6df8a0addc9a37b48c5a1827dd3f65f2ed57cf), and based on that I suspect that checking for `no basic auth credentials` was merely a convenience (pick a string in the output that may be a bit more stable).


The first one (`TestLogoutWithExternalAuth`) looks to be using the error-message to try to detect failures after a logout. There's a _LOT_ in that test that is testing CLI behavior (config-file location, log-in, log-out), and it really feels like those parts would make a lot more sense in the docker/cli repository; this (moby) repository on the other hand could benefit from an API test for the same, and to verify that the API returns the expected response and status-codes;
https://github.com/moby/moby/blob/f6576e9a741d8ee9152a224d2ef342decdcab93c/integration-cli/docker_cli_logout_test.go#L60-L63

The second one (`TestPushNoCredentialsNoRetry`) may be even more problematic; it's testing behavior "by proxy"; that the CLI does _NOT_ print `Retrying` (with a capital `R`); https://github.com/moby/moby/blob/f6576e9a741d8ee9152a224d2ef342decdcab93c/integration-cli/docker_cli_push_test.go#L219-L220


I suspect we can make that test pass by changing `Retrying` to lowercase in the progress :see_no_evil: so we should probably consider having a test that verifies that we don't retry (not that we don't _print_ that we're retrying) https://github.com/moby/moby/blob/f6576e9a741d8ee9152a224d2ef342decdcab93c/distribution/xfer/download.go#L302

Reply:
Just checking for `ErrInvalidAuthorization` is not enough because it's also returned for example when you try to pull a non-existing repository, or if the credentials are provided, but do not give access to that repository.

If we'd decorate every `ErrInvalidAuthorization`, we'd end up with something like this:
```
Error response from daemon: pull access denied for aodkoakds, repository does not exist or may require 'docker login': failed to resolve reference "[docker.io/library/aodkoakds:latest](http://docker.io/library/aodkoakds:latest)": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

Reply:
But, perhaps we don't really need to have an explicit `no basic auth credentials` in the message and the `pull access denied for %s, repository does not exist or may require 'docker login'` would be sufficient?

Reply:
> because it's also returned for example when you try to pull a non-existing repository, or if the credentials are provided, but do not give access to that repository.

Hm, right, but that's roughly what the error message describes; "image doesn't exist" (non-existing repository), or "may require 'docker login'".

The _"credentials are provided, but do not give access to that repository"_ is indeed confusing if we show "may require `docker login`", but that's the part that's addressed in the containerd PR (so we need to wait for that one to land?)

I'm somewhat wondering if we should wrap the error message at all, or return as-is; we still (at least for existing API version) would have to change the response-code to be a 404 (something we should consider changing), and of course make sure that no retry happens (I _think_ that may be handled in containerd's code?), but looking at your example;


```
1. Error response from daemon: 
2.  pull access denied for aodkoakds, repository does not exist or may require 'docker login': 
3.    failed to resolve reference "docker.io/library/aodkoakds:latest": 
4.      pull access denied, repository does not exist or may require authorization: 
5.        server message:
6.           insufficient_scope:
7.             authorization failed
```

`1.` is added by the CLI
`2.` is wrapping we're adding here, but the same information is already captured in containerd's error
`6.` is (I _think_?) response from the registry, so out of our hands (I'd love the "server message" to somehow indicated "server means registry here" though :thinking:)
`7.` is (unfortunately) needed in containerd's way of coding errors (for users it's mostly noise)

From the above, I _think_ we can safely remove `2.` if we're ok with losing the `docker login` part (I guess _ideally_ we'd get rid of the "convert to NotFound" part, so that the CLI could add this if we want to?)

Reply:
Dropped the wrapping